soroban-env-host/src/native_contract/account_contract.rs

```diff
@@ -0,0 +1,234 @@
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub enum Signature {
+    Ed25519(BytesN<64>),
+    Account(HostVec),
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct AccountEd25519Signature {
+    pub public_key: BytesN<32>,
+    pub signature: BytesN<64>,
+}
+
+impl AuthorizationContext {
+    fn from_invocation(host: &Host, invocation: &AuthorizedInvocation) -> Result<Self, HostError> {
+        let top_invocation = invocation
+            .call_stack
+            .last()
+            .ok_or_else(|| host.err_general("empty auth call stack"))?;
+        let args = HostVec::try_from_val(
+            host,
+            host.scvals_to_rawvals(invocation.top_args.0.as_slice())?,
+        )?;
+        Ok(Self {
+            contract: BytesN::try_from_val(
+                host,
+                host.bytes_new_from_slice(top_invocation.contract_id.0.as_slice())?,
+            )?,
+            fn_name: top_invocation.function_name,
+            args,
+        })
+    }
+}
+
+pub(crate) fn check_generic_account_auth(
+    host: &Host,
+    generic_account_contract: &Hash,
+    signature_payload: &[u8; 32],
+    signature_args: Vec<RawVal>,
+    invocations: &Vec<AuthorizedInvocation>,
+) -> Result<(), HostError> {
+    let payload_obj = host.bytes_new_from_slice(signature_payload)?;
+    let signature_args_vec = HostVec::try_from_val(host, signature_args)?;
+    let mut auth_context_vec = HostVec::new(host)?;
+    for invocation in invocations {
+        auth_context_vec.push(AuthorizationContext::from_invocation(host, invocation)?)?;
+    }
+    // TODO: auth_context_vec now contains account which may be problematic from
+    // the security and safety perspective. Probably AuthorizationManager and Host
+    // should fail on any account access while checking account auth (one more
+    // reason for the special mode for the check_auth invocations).
+    Ok(host
+        .call_n_internal(
+            generic_account_contract,
+            Symbol::from_str("check_auth"),
+            &[
+                payload_obj.into(),
+                signature_args_vec.into(),
+                auth_context_vec.into(),
+            ],
+            // Allow reentry for this function in order to do wallet admin ops
+            // within the auth framework. Maybe there is a more elegant way 
+            // around this.
+            // TODO: check if there are security concerns about this.
+            true,
+        )?
+        .try_into()?)
+}
+
+pub(crate) fn check_account_authentication(
+    host: &Host,
+    account_id: &ScAccountId,
+    payload: &[u8],
+    signature_args: Vec<RawVal>,
+) -> Result<(), HostError> {
+    if signature_args.len() != 1 {
+        return Err(err!(
+            host,
+            ContractError::AuthenticationError,
+            "incorrect number of signature args: {} != 1",
+            signature_args.len() as u32
+        ));
+    }
+    let signature: Signature = signature_args[0].try_into_val(host).map_err(|_| {
```

Comment:
It looks like a `Signature` in the accounts case is actually a vector of `AccountEd25519Signature`, which is a little confusing. Would it be possible to replace `Signature` with `AccountEd25519Signature` and use a vector of that for both the classic account and ed25519 case? I guess you'd unnecessarily pass in the public key in the signature for Ed25519, but it sounds like we might remove that Identifier type anyways?